kork-retrofit/src/main/java/com/netflix/spinnaker/kork/retrofit/ErrorHandlingExecutorCallAdapterFactory.java

```
@@ -0,0 +1,279 @@
+
+/*
+ * {@link retrofit.RetrofitError} and {@link retrofit.ErrorHandler} are no longer present in retrofit2.
+ * So this class helps to achieve similar logic as retrofit and handle exceptions globally in retrofit2.
+ * This can be achieved by setting this class as CallAdapterFactory at the time of {@link Retrofit} client creation.
+ *  */
+public class ErrorHandlingExecutorCallAdapterFactory extends CallAdapter.Factory {
+
+  /*
+   * Which need to be set only if clients uses async call i.e enqueue method.
+   * Clients which make use of sync call i.e execute method only, no need to set this.
+   * */
+  private final @Nullable Executor callbackExecutor;
+
+  ErrorHandlingExecutorCallAdapterFactory() {
+    this.callbackExecutor = null;
+  }
+
+  ErrorHandlingExecutorCallAdapterFactory(Executor callbackExecutor) {
+    this.callbackExecutor = callbackExecutor;
+  }
+
+  public static ErrorHandlingExecutorCallAdapterFactory getInstance(Executor callbackExecutor) {
+    return new ErrorHandlingExecutorCallAdapterFactory(callbackExecutor);
+  }
+
+  public static ErrorHandlingExecutorCallAdapterFactory getInstance() {
+    return new ErrorHandlingExecutorCallAdapterFactory();
+  }
+
+  /*
+   * Returns a call adapter for interface methods that return {@code returnType}, or null if returnType is not instance of Call.
+   *
+   * throws IllegalArgumentException if returnType is not ParameterizedType.
+   * */
+  @Nullable
+  @Override
+  public CallAdapter<?, ?> get(Type returnType, Annotation[] annotations, Retrofit retrofit) {
+
+    /*
+     * Expected the raw class type from returnType to be Call class
+     * otherwise return null as it cannot be handled by this factory
+     * */
+    if (getRawType(returnType) != Call.class) {
+      return null;
+    }
+
+    /*
+     * check whether returnType is ParameterizedType, else throws IllegalArgumentException.
+     * */
+    if (!(returnType instanceof ParameterizedType)) {
+      throw new IllegalArgumentException(
```

Could the comment here explain why we're throwing an IllegalArgumentException instead of returning null since the above code returns null and the javadoc for the method we're implementing says to return null....or maybe switch to returning null here too?

If they're not too contrived...like they actually could happen because someone is trying to use retrofit2 properly, but doesn't understand it, I'd like to see tests for these two failure paths....or at least comments in the code that explain how these paths can happen, ideally with examples.

If the tests "artificially" call this method with specially-crafted params, as opposed to something close to a "real" usage, let's not do it.

---

as per discussion, it is better to return null than throw an exception, so returning null